Cython/Utility/Overflow.c

```diff
@@ -95,20 +108,28 @@ static CYTHON_INLINE {{UINT}} __Pyx_mul_{{NAME}}_checking_overflow({{UINT}} a, {
         return r;
 #endif
     } else {
-        {{UINT}} prod = a * b;
-        double dprod = ((double) a) * ((double) b);
+        // somewhat surprisingly, at least on x86_64 it's quicker to use floats (c.f. the const
+        // variant below)
+        {{UINT}} r = a * b;
+        float fprod = ((float) a) * ((float) b);
         // Overflow results in an error of at least 2^sizeof(UINT),
-        // whereas rounding represents an error on the order of 2^(sizeof(UINT)-53).
-        *overflow |= fabs(dprod - prod) > (__PYX_MAX({{UINT}}) / 2);
-        return prod;
+        // whereas rounding represents an error on the order of 2^(sizeof(UINT)-23).
+        *overflow |= fabs(fprod - r) > ((float)(__PYX_MAX({{UINT}})) / 2);
+        return r;
     }
 }
 
 static CYTHON_INLINE {{UINT}} __Pyx_mul_const_{{NAME}}_checking_overflow({{UINT}} a, {{UINT}} b, int *overflow) {
-    if (b > 1) {
-        *overflow |= a > __PYX_MAX({{UINT}}) / b;
+    if (__Pyx_constant_p(a) && !__Pyx_constant_p(b)) {
+        // paranoia
```

Comment:
I'm aware that there's a longer comment in the signed variant further down, but it seems worth repeating it here. For a future reader, it will be less obvious what the 'paranoia' refers to. (I'm ok with keeping this safety check in general.)